### ISSUE TITLE
Remove onSuccess and onError

### ISSUE THREAD

**Issue by egeriis** [dixieio/redux-json-api#48]
These methods were deprecated as of #17. Remove the callbacks in v2.0.

**Comment by kazazor** [comment]
Quick question. We would like to start using this package for our development. Thank you for that :)

My question, why do you use the Promise like syntax for your dispatch? It feels a little bit not redux way.. I would expect it to work more like the [redux-api-middleware](https://github.com/agraboso/redux-api-middleware) and dispatch FAILURE / SUCCESS instead of me having to control it using `then` / `catch`.. Are there any plans on supporting something like that?

**Comment by egeriis** [comment]
Hi @kazazor

Happy to hear that you're considering to use this library. 

I'm not sure if the techniques from redux-api-middleware can be applied to redux-json-api. But we're still learning Redux, so we're very open to suggestions 🙂

I'll have a more thorough look at redux-api-middleware tomorrow.

**Comment by kazazor** [comment]
Thank you :)
In general, the idea is that everything goes through the `dispatch`.. Even the failures and successes. So there is no need for any use of `then` / `catch`. Instead, if for example if dispatching the action GET_ALL_ITEMS, then the middleware will dispatch **REQUEST**_GET_ALL_ITEMS, **SUCCESS**_GET_ALL_ITEMS & **FAILURE**_GET_ALL_ITEMS when needed.

Currently what we're looking on doing is wrap the `then` / `catch` you're providing and dispatch the FAILURE / SUCCESS when needed. That way we'll be 100% redux way.
Of course it would be great not to do that and just have it built in :)
We did this kind of behavior in another inner package we developed as a middleware inside our organization. Should be really straight forward.

Would love to hear what you think once you'll have a look at redux-api-middleware 😄

**Comment by egeriis** [comment]
Hmm. We're already dispatching events in case of success/failure. For `readEndpoint` is is `API_READ` (success) and `API_READ_FAILED` (fail/error).

How would you tap into this? A custom extension of the API namespace?

**Comment by OmriAharon** [comment]
Hey @egeriis, at the moment what we do is create middleware ourselves that translate said actions to new actions that are more descriptive using the requested endpoint string. A sort of a middleware parser, if you will.

 I'd like to suggest a way to tackle this in a way that would be easier to control and maintain.
At the moment, like you said, we get actions like `API_READ` dispatched upon completion. But since it's too generic, all reducers get that and have to check if it's their data that was received, or was triggered by a different endpoint call. 

If there was a way, such as in the API middleware of redux, to pass in an object of actions to trigger that would make things much easier. E.g.:
```

dispatch(readEndpoint('tasks', {
    onSuccessAction: 'TASKS_READ',
    onFailAction: 'TASKS_READ_FAILURE'
}));
```

**Comment by egeriis** [comment]
@OmriAharon I see your point. But there's actually an alternative approach, as far as I see it, which doesn't clutter the API of _redux-json-api_:

```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint('tasks'))
    .then(tasksRead)
    .catch(tasksFailedToRead);
};
```

Wouldn't that work as flexible?

**Comment by OmriAharon** [comment]
That would be flexible, yes. 

However, returning a promise doesn't preclude the object suggested above.
I can see both approaches living side by side. 
That way, the developers can choose for themselves.

What do you think?

**Comment by egeriis** [comment]
I don't believe that the best solution is multiple solutions. I would also go far to keep a simple API, as long as it doesn't prevent obvious use cases and implementations.

This library should be flexible enough to fit 80% of use cases, but simple enough that it doesn't weigh too much in the application.

**Comment by kazazor** [comment]
@egeriis - I agree that multiple solutions could be tricky.
However, if we'll have to choose 1 way I would choose @OmriAharon way.

* Adding it is very very simple (you're already doing a dispatch of an action, so just let me tell you which to dispatch).
* You're already doing a dispatch along with the Promise.. So that kind of contradicts with your suggestions..
* Promise like is not really a redux look and feel solution. That's why when  started using the library that wasn't the straight forward solution.

What do you think?

**Comment by egeriis** [comment]
My point is that the API should be as simple as possible. The suggestion @OmriAharon proposed would add to the API, thus making it more complex.

I don't know why you're considering whether it's a Redux best practice, because Redux is a data store and doesn't care about anything but state-altering actions. Asynchronous action dispatching is a redux-thunk feature, and they even provide an example of the method we've implemented in their docs: https://github.com/gaearon/redux-thunk#composition (the ability to chain on a returned Promise).

When there is a solution, I don't want to pollute the API. That's reasonable and rational. What you're suggesting is opinionated and taste, which makes it completely irrational to demand that that is the only right way. I hope that makes sense 🙂

**Comment by OmriAharon** [comment]
The only thing I find weird is that an action will be dispatched (API_READ, etc..) regardless of the promise implementation, and that there is absolutely no way to modify the action name.

I've created a middleware to bypass that, I'm sure most others would do the same. It's reasonable to assume that the developer would have to create a new action to dispatch in the `then` resolution to store the result in redux anyway, why not save him/her the trouble?

**Comment by egeriis** [comment]
`API_WILL_READ` and `API_READ` are there for a purpose. So are the equivalent for the rest of the CRUD methods. Their purpose is outlined in the documentation and I don't understand why anybody would disable them.

Adding what you're suggesting would dispatch _another_ action. I can't explain my stance better than the last paragraph of my previous comment. 

I haven't been provided with any argument(s) suggesting that the current approach doesn't work in every use case. The only argument I've seen is "I don't like this approach".

**Comment by kazazor** [comment]
The arguments are very simple:

1. Why do you force in your API to **always** use the `then` & `catch` & `dispatch`? We can always use that approach that you're suggesting, but why do I need to duplicate this code over and over again? Supplying your code with the action to dispatch will reduce the developer's code because you'll be saving the trouble for others to write that over and over again.

2. The `API_WILL_READ` is being dispatched for **every** request we're making, why do I need all of these actions with no reference on what they are being dispatched? Even for logging, that redux actions gives us so beautifully, this approach only makes a lot of noise that doesn't tell me anything.
What if I have X requests at the same time? I'll get X `API_WILL_READ` that only makes things confusing and hard to maintain.
For example, having `API_READ_TASKS` / `API_READ_LISTS` or actions like that I'm supplying will make everything much more readable.
As a developer that uses the package, I would expect to tell you what to dispatch in my system.. Otherwise it is just background noise..

Do you agree with me that these reasons are not just "I don't like this approach"?

**Comment by egeriis** [comment]
1. That is not an argument, that is a question. Which I believe I have addressed extensively. Give me an example where you will have duplicate code, and how that would potentially look with your suggested addition? Because I haven't understood correctly, if your suggestion isn't as much duplication as status quo.
2. That's a completely different discussion, with a very valid question: does the functionality provided by dispatching `API_WILL_READ` and `API_READ` provide any value. If this is a discussion that you would like to extend, it would make sense to have a separate thread for that. That thread should include your proposal for what to do about these actions.

**Comment by kazazor** [comment]
Regarding number 2, this is the same discussion.
Have you looked at the [redux-api-middleware](https://www.npmjs.com/package/redux-api-middleware)? They're using the **exact same approach we're suggesting**.

If you'll provide the action to dispatch you will solve number 2 and will make the actions actually meaningful.
By doing that you'll also save some code duplication like this:

Your approach:
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint('tasks'))
    .then((data) => {
         dispatch({
                 type: 'TASKS_SUCCESS', 
                 payload: data
        });
     })
    .catch((error) => {
          dispatch({
                 type: 'TASKS_FAILED',
                 payload: error
          });
     });
};

const createTask = () => dispatch => {
  dispatch(readEndpoint('createTask'))
    .then((data) => {
         dispatch({
                 type: 'CREATE_TASKS_SUCCESS', 
                 payload: data
        });
     })
    .catch((error) => {
          dispatch({
                 type: 'CREATE_TASKS_FAILED',
                 payload: error
          });
     });
};
```

The suggestion (option number 1 with another parameter):
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint('tasks', 'TASKS'));
};

const createTask = () => dispatch => {
  dispatch(readEndpoint('createTasks', 'CREATE_TASKS'));
};
```

The suggestion (option number 2 with an options parameter):
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint({ 
          endpoint: 'tasks',
          dispatchAction:  'TASKS' 
  }));
};

const createTask = () => dispatch => {
  dispatch(readEndpoint({ 
          endpoint: 'createTasks', 
          dispatchAction:  'CREATE_TASKS' 
  }));
};
```

In both options you will add a suffix to the action name according to its type of dispatch.
So for example, the dispatched actions will be:

* `TASKS_READ` & `CREATE_TASKS_READ` when making the calls.
* `TASKS_SUCCESS` & `CREATE_TASKS_SUCCESS` when the call succeed - with the payload data.
* `TASKS_FAILED` & `CREATE_TASKS_FAILED` if the call fails - with the payload error.

# The best option IMO
Another option which is more straight forward will be to also provide the exactly what actions to dispatch for each option (read, success & failed):
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint({ 
         endpoint: 'tasks', 
          readAction: 'TASKS_READ', 
          successAction: 'TASKS_SCCUESS', 
          failedAction:  'TASKS_FAILED' 
  }));
};

const createTask = () => dispatch => {
  dispatch(readEndpoint({ 
          endpoint: 'createTasks', 
          readAction: 'CREATE_TASKS_READ', 
          successAction: 'CREATE_TASKS_SCCUESS', 
          failedAction:  'CREATE_TASKS_FAILED' 
  }));
};
```

Or even shorter like api middleware are doing, provide an array for the actions in the format `[READ, SUCCESS, FAILURE]`:
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint({ 
         endpoint: 'tasks', 
         actionTypes: ['TASKS_READ', 'TASKS_SCCUESS', 'TASKS_FAILED' ]
  }));
};

const createTask = () => dispatch => {
  dispatch(readEndpoint({ 
          endpoint: 'createTasks', 
          actionTypes: ['CREATE_TASKS_READ', 'CREATE_TASKS_SCCUESS', 'CREATE_TASKS_FAILED' ]
  }));
};
```
In this option, you can even control **what actions to dispatch if at all**. So, if I'll not provide you with a `readAction`, you won't dispatch it. That means I didn't want it, please let me control that need by myself.

<hr>

As you can see, I do see it as the same discussion.
<h4>It gives me more options to control with less code.</h4>
I don't need to do `then`, `catch` & `dispatch` inside of them all the time. You're wrapping that and doing that for me instead.
Why not hit two birds in one stone?

<hr>

I believe I addressed the suggestion extensively, let me know what you think.

**Comment by egeriis** [comment]
> Regarding number 2, this is the same discussion.
Have you looked at the redux-api-middleware? They're using the exact same approach we're suggesting.
>
> If you'll provide the action to dispatch you will solve number 2 and will make the actions actually meaningful.

It's a bit frustrating that it seems like you're not reading my replies thoroughly. As I've written, `API_WILL_READ` and `API_READ` provides functionality. They're not just dummy actions dispatched for fun and giggles.

**Comment by kazazor** [comment]
@egeriis - I can assure you I feel even worse than you.
I came into your repo trying to help and suggest more options an opinions - this is not even my code.

I even wrote you a LOT of code examples showing you the different options, and it doesn't seem like you even give a damn.

I can tell you that as an open source project that is the worse feeling you as an owner can give.
After that big code example you really haven't looked at it at all?!

I give up, good luck with the package.

**Comment by egeriis** [comment]
I am writing a reply regarding the code. I decided to split it up into two comments. I am attempting to be respectful, but I feel like we're missing each other. Getting some code to talk about was my attempt to become more concrete in the discussion. Hopefully we can achieve that. Give me a few minutes to reply.

**Comment by egeriis** [comment]
Regarding the code:

I'm not sure how _redux-json-api_ would dispatch an application specific action based on a string passed to e.g. `readEndpoint`. Example:

```js
dispatch(readEndpoint({ 
  endpoint: 'tasks', 
  readAction: 'TASKS_READ'
}));
```

What do you expect _redux-json-api_ to do with "TASKS_READ"?

**Comment by kazazor** [comment]
>I am attempting to be respectful

I'm sorry, but this whole discussion is not respectful.
I'm reading ALL of your replies and writing BIG code example to YOUR repo in my private time.
If you see that you didn't understand me I would even do Skype chat with you to help, but assuming that I'm not reading your comments just because you didn't fully understand me is highly disrespectful.

<hr>

My expectations are as followed:

1. redux-json-api makes an API call
* It dispatches the "TASKS_READ" action
* The requests returns with success / error
* It dispatches the "TASKS_SUCCESS" / "TASKS_FAILED" (depends on the response)

It does that exactly like you're doing that, but instead don't return a promise and just in the package itself dispatch according the return from the API..

As you saw that is less code with more options for the developer and less code.
Like in the code example I wrote above:
```js
const readTasks = () => dispatch => {
  dispatch(readEndpoint({ 
         endpoint: 'tasks', 
          readAction: 'TASKS_READ', 
          successAction: 'TASKS_SCCUESS', 
          failedAction:  'TASKS_FAILED' 
  }));
};

const createTask = () => dispatch => {
  dispatch(readEndpoint({ 
          endpoint: 'createTasks', 
          readAction: 'CREATE_TASKS_READ', 
          successAction: 'CREATE_TASKS_SCCUESS', 
          failedAction:  'CREATE_TASKS_FAILED' 
  }));
};
```

**Comment by egeriis** [comment]
> I'm sorry, but this whole discussion is not respectful.
> I'm reading ALL of your replies and writing BIG code example to YOUR repo in my private time.

Sorry that you feel that way. I might be coming out a bit frustrated. Allow me to apologize and make up for that.

--

Since an action is not simply a string, how would "TASKS_READ" be dispatched? I can't see how _redux-json-api_ would find the action based on a string. Could you elaborate?

**Comment by egeriis** [comment]
Wait, I think I understand what you're looking for with that string. Is it correctly understood that _redux-json-api_ would then dispatch the following action:

```json
{
  "type": "TASKS_READ",
  "payload": responseData
}
```

Where `responseData` is the JSON API response from the API?

**Comment by kazazor** [comment]
Yes that is correct 🎉 
You'll get all of the pros I wrote [in this comment](https://github.com/dixieio/redux-json-api/issues/48#issuecomment-258694434) just by doing that.

* Less code
* More control to the developers
* Less noise of dispatched actions with no meaning

**Comment by egeriis** [comment]
> Less code

This is probably true, as you can see in my elaboration below.

> More control to the developers

Can you elaborate on this?

> Less noise of dispatched actions with no meaning

This is actually false. `API_WILL_READ` and `API_READ` will still be dispatched as outlined earlier. So it will add another action to the stack.

--

So I decided to outline full examples for acting on the response body using the current solution and with the configuration method you're proposing.

# Promise vs action

In this example I am requested from an endpoint which returns a resource collection of type "users". 

## Using Promise

Here is an application implementation using _redux-json-api_ in it's current form:

This is the action creators:

```js
const SAVE_USER_IDS = 'SAVE_USER_IDS';
const readUsersAndSaveIds = () => dispatch => {
  dispatch(readEndpoint('users'))
    then(response => dispatch({
      type: SAVE_USER_IDS,
      userIds: response.data.map(user => user.id)
    }))
};

const reducer = (state, action) => {
  switch (action.type) {
    case SAVE_USER_IDS:
      return Object.assign({}, state, {
        userIds: action.userIds
      });
    default:
      return state;
  }
}
```

And the action would be triggered like this:

```js
dispatch(readUsersAndSaveIds());
```

## Using a configuration

Here the action will be dispatched with a `payload` property that will contain the full response body.

```js
const reducer = (state, action) => {
  switch (action.type) {
    case 'USERS_READ':
      return Object.assign({}, state, {
        userIds: action.payload.data.map(user => user.id)
      });
    default:
      return state;
  }
}
```

```js
dispatch(readEndpoint('users', { readAction: 'USERS_READ' }));
```

## Conclusion

After doing this exercise it still appears like a matter of taste, which solution is preferred. You may like the latter, but I personally prefer the first option.

--

What are you thoughts on this?

**Comment by egeriis** [comment]
Let me just add the point that it's not the purpose that you should do this for every API call. _redux-json-api_ is supposed to have you covered for 80% of your API calls, without having to do anything but e.g. `dispatch(readEndpoint("users"))`.

**Comment by kazazor** [comment]
> Let me just add the point that it's not the purpose that you should do this for every API call. redux-json-api is supposed to have you covered for 80% of your API calls, without having to do anything but e.g. dispatch(readEndpoint("users")).

Yes this is correct, in my examples just think of it as `tasks` / `lists` / `members` etc.. Didn't mean to show `tasks` & `createTask` endpoints.. Let's assume we have different entities here..
<hr>
So we have:
1. Less code - ✔️ 

> "Less noise of dispatched actions with no meaning"
This is actually false. API_WILL_READ and API_READ will still be dispatched as outlined earlier. So it will add another action to the stack.

I meant to have the actions to dispatch **instead** of the `API_WILL_READ` & `API_READ`.. Not in addition. So it is actually true - it will be less noisy.

Why less noisy?
Let's assume you're making multiple calls to different end points (again, `tasks` / `lists` / `members` etc..) at the same time (an example for that is an initialization code for your app that go and gets a lot of data).
What you'll start to see is `API_WILL_READ` & `API_READ` being throw all over the place. What can it actually tell me? Except for just being a line in my console if I log it?
To actually know and **see the flow of my app calls** I'll need to check the payload for specific things **that might change over time** to know from which API it was called..
Instead - changing the dispatched actions to **only** use what was provided by the user will always give you meaningful actions being dispatched.
That is why it is less noisy - ✔️

> "More control to the developers"
Can you elaborate on this?

Like I said before:
> In this option, you can even control what actions to dispatch if at all. So, if I'll not provide you with a readAction, you won't dispatch it. That means I didn't want it, please let me control that need by myself.

So I as a developer can decide if I don't care of the READ action to be thrown, I can tell you not to do that. So now I'll only see for that endpoint only the success and failed actions being thrown.
It is more control for me and could even be less noise if I choose to remove some actions along the way.
Maybe you didn't have that use case so far but other might have that. So that gives me more control - ✔️

<hr>
# Conclusion
In your examples you compared between the two options and your conclusion was that the other option results with less code, and I think you can agree that in a redux world app you don't lose readability here since this is how most of your app data works.
Even just that could mean to go and make it happen.

I hope that I also showed you not just 1 pro for that, but total of 3 now.

>After doing this exercise it still appears like a matter of taste, which solution is preferred. You may like the latter, but I personally prefer the first option.

I still haven't seen any reason from your end when is the `then` & `catch` method is better.
Given that, if you narrowed it down to just preferred way - I think I showed why this is also gives your more options with less code and less noise.

Also, something important - you don't have to kill any option! (well IMO you have to kill the `API_WILL_READ` & `API_READ`.. Just causes noise..)
You can keep the Promise like for situations it is better, and also support the dispatch by given actions.. It lives together and will give the package much more use cases and strength.

**Comment by egeriis** [comment]
I have a few comments:

> I meant to have the actions to dispatch instead of the API_WILL_READ & API_READ.. Not in addition. So it is actually true - it will be less noisy.

Then you're removing functionality from the library. Removing `API_READ` will actually remove the whole idea behind this lib. This is what makes me frustrated in this discussion, because it appears you haven't read the documentation since it seems that you don't know what `API_WILL_READ` and `API_READ` does.

> What you'll start to see is API_WILL_READ & API_READ being throw all over the place. What can it actually tell me? Except for just being a line in my console if I log it?

Actions aren't supposed to tell you stuff, they're supposed to change state. And both `API_WILL_READ` and `API_READ` does.

> In this option, you can even control what actions to dispatch if at all. So, if I'll not provide you with a readAction, you won't dispatch it. That means I didn't want it, please let me control that need by myself.

It appears that you're mixing `API_READ` with acting on the promise, e.g. through `then`. Your `readAction` is not the equivalent of `API_READ`. So this flexibility is already in place, since it's up to you whether you'd like to add a promise resolver or catch a promise error.

> Also, something important - you don't have to kill any option!

I know, but just to be clear, we're not going to have two ways of achieving the same result.

**Comment by kazazor** [comment]
> Actions aren't supposed to tell you stuff, they're supposed to change state. And both API_WILL_READ and API_READ does.

Actions are suppose to **also** give you information on what's going on in your system.
That is why projects like [redux logger](https://github.com/evgenyrodionov/redux-logger) are so popular in dev env.. You can say that you don't use that in your development (which is a shame IMO) but this whole pattern of actions was also designed for better development and code maintenance.

>because it appears you haven't read the documentation since it seems that you don't know what API_WILL_READ and API_READ does.

The `API_READ` & `API_WILL_READ` was just an example. I did read all of your [documentation](https://github.com/dixieio/redux-json-api/blob/master/docs/api.md#api-methods) and the following links inside of it.

But I got sick of you disrespecting me and assuming that I do not know what I'm talking about, when in reality it seems that no matter how many times I showed you redux api middleware you didn't even went inside once to see the code implementation they did - which could have saved us a lot of time here. Just 1 look at the implementation and usage and you would have understand everything and we could have finished that much sooner.

I think there is not point to discuss that anymore. We've already created a wrapper for that instead.. We just wanted to help out here but it seems that it won't happen since you have lost (for no good reason) any trust and you're not looking at our point here.

I wish you all the best.

**Comment by egeriis** [comment]
@kazazor Look, I really appreciate that you're coming to us to propose a change in our library. But we cannot have a discussion where I have to repeat myself. I know that you feel the same, but we actually have a pretty good dialogue before you bailed on the discussion. 

You keep iterating that we can just remove `API_WILL_READ` and `API_READ`, but that is just not possible. Well, we can remove `API_WILL_READ`, but that will also remove the functionality that it implements. How is that just an example? What other examples do you have?

I really wish to learn about your use case. And yes I require you to be detailed, precise and willing to have a discussion, if you wish to be heard.

**Comment by kazazor** [comment]
I'm sorry @egeriis, but I don't see a partner here to continue the discussions and you currently feel the same.

I don't believe it is worth your or my time nor it will make any difference in this situation.
No hard feelings, the package works - it can stay like that :)

**Comment by egeriis** [comment]
I surely don't feel that way, @kazazor. But I respect that you do not wish to carry on. Thank you for using our library and please don't hesitate to get back if you wish to continue this discussion or have any other feedback.

**Comment by nitzansan** [comment]
We're also facing an identical issue.
I would have to agree with @kazazor here - it would be extremely(!) convenient to have the ability to specify action names for the different endpoints methods.

We have actually used promises at first, since the current one action name for which endpoint you're calling is not really helpful for handling events in a redux-way.
This caused a lot of code duplication as well as a lot of maintenance and time to develop.
Eventually we had to create a reducer that listens to all the redux-api-json events, parses the payloads and then triggers relevant actions based on that. Also this approach has its own limitation.

It would be great if we could specify that action names for the methods as described here by kazazor.
I don't think it matters at all if there are two actions being dispatched, both the "native" current actions (e.g API_WILL_READ) followed by a custom specified action (e.g PERFORMANCE_DETAILS_WILL_READ), both with the same payloads.

Otherwise, if the current "native" actions can be implemented without fetching their actions in some way, then there can be just the custom action dispatched, but if it's tightly coupled then I think people would not mind at all having two actions dispatched.
It's actually what happens anyways with the workarounds around this problem.

**Comment by egeriis** [comment]
@nitzansan I'd be happy to take up the discussion, but please provide your use case for me to get an insight. Otherwise it's very hard for me to get into your perspective. That was the challenge for me in the discussion above.

It sounds like what you're looking for is a way to _structurally_ chain actions with the redux-json-api actions? 

What about doing it in your own module, like this:

```js
import { readEndpoint } from 'redux-json-api';

const API_DID_READ_USERS = 'API_DID_READ_USERS';

const didReadUsers = (payload) => ({ type: API_DID_READ_USERS, payload });

const readUsers = () => (dispatch, getState) => { 
  readEndpoint('/users')
    .then(res => dispatch(didReadUsers(res))
};

const reducer = (state, action) => {
  switch(action.type) {
    case API_DID_READ_USERS:
      console.log(action.payload);
      // update state
      return state;
    default:
      return state;
  }
};
```

**Comment by nitzansan** [comment]
@egeriis - What you gave is a very good specific workaround which we actually implemented around our code.
The use case is quite plain to tell you the truth - we want the ability to trigger custom dispatches based on different usages from different components and logics.

There are many ways to achieve this. Either by using promises, wrapping the readEndpoint function, using middleware.
However, it seems that this custom action is something that many users will want to have and probably many users has been handling and work-arounding with this.
I know that another team in our office also started to use this package and they also struggle with ways to customize the actions.

While there are many ways to achieve this goal, I think that the best one is actually transparent to the consumers of this package - providing an easy interface through the various APIs.
for example:
readEndpoint({ 
         endpoint: 'users', 
          initiatedAction: 'API_USERS_WILL_READ', 
          successAction: 'API_USERS_READ_SUCCESS', 
          failedAction:  'API_USERS_READ_FAILED' 
  })
deleteEntity({ 
         endpoint: 'users/1', 
          initiatedAction: 'API_USER_WILL_DELETE', 
          successAction: 'API_USER_DELETED', 
          failedAction:  'API_USER_DELETE_FAILED' 
  })

This way, the actual implementation will be under-the-hood and the interface should be very easy to use for all the users ( + it's very easy to understand what actions you expect to be triggered just from glancing since it's a plain object and not a middleware/promise/wrapper that you need to traverse to and actually read :neckbeard: )

I didn't go yet too much into the code of this project, so I don't know if the current actions can be emitted in case of a custom action given, but even if not I don't there should be any issues with dispatching two actions (API_WILL_READ + API_USERS_WILL_READ for example).

Hope I was coherent enough. Let me know if anything is not clear. 😄

**Comment by egeriis** [comment]
There will be no issue with this approach 🙂

I honestly just don't agree that this approach is better. But that is obviously very opinionated, and some of the thoughts that arise are really about application code structure. Obviously I do not want to discuss how other people should structure their apps.

I'm not sure what I personally would like to do from here. The easy solution would be to cave in, but there's obviously always a need to avoid feature creep when building software. 

When building and maintaining _redux-json-api_, we focus on keeping the API simple, while providing all the flexibility that is needed to support most applications. Adding these custom actions would _add_ to the API, thus making it more complex, and would not add any extra flexibility. The hard thing in this discussion, is that it's a matter of taste and code structure, as I see it.

Btw. do you have a use case example? Maybe that would be useful :)

**Comment by Nopzen** [comment]
Merged to branch 2.0 and is in the progress of release. Closing.
